In [5]:
!pip install ipython-sql pymysql

## Weaving the data threads of Maji Ndogo's narrative
### Maji Ndogo: From analysis to action

In [3]:
%reload_ext sql
%config SqlMagic.style = '_DEPRECATED_DEFAULT'
%config SqlMagic.autopandas = True

In [4]:
%sql mysql+pymysql://root:12345678@localhost:3306/md_water_services

### Integrating the Auditor's report

In [6]:
%%sql
SELECT*
FROM auditor_report
LIMIT 10;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
10 rows affected.


location_id type_of_water_source  true_water_source_score  \
0   SoRu34980                 well                        1   
1   AkRu08112                 well                        3   
2   AkLu02044                river                        0   
3   AkHa00421                 well                        3   
4   SoRu35221                river                        0   
5   HaAm16170                 well                        1   
6   AkRu04812                 well                        3   
7   AkRu08304                 well                        3   
8   AkRu05107                 well                        2   
9   AkRu05215                 well                        3   

                                          statements  
0  Residents admired the official's commitment to...  
1  Villagers spoke highly of the official's dedic...  
2  Villagers were touched by the official's inter...  
3  Villagers were moved by the official's visit, ...  
4  A photographer's lens captures the queue, thou...  
5  With an open heart, the official created an at...  
6  The official's presence left an indelible mark...  
7  The official's interactions resonated deeply w...  
8  Villagers spoke highly of the official's dedic...  
9  Villagers admired the official's visit for its...

1. Is there a difference in the scores?
2. If so, are there patterns?

For the first question, we will have to compare the quality scores in the water_quality table to the auditor's scores. The auditor_report table
used location_id, but the quality scores table only has a record_id we can use. The visits table links location_id and record_id, so we
can link the auditor_report table and water_quality using the visits table.

In [11]:
%%sql
SELECT 
ar.location_id AS audit_location,
ar.type_of_water_source,
v.location_id AS visit_location,
v.record_id
FROM md_water_services.auditor_report AS ar
JOIN md_water_services.visits AS v
ON ar.location_id=v.location_id
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


audit_location type_of_water_source visit_location  record_id
0      SoRu34980                 well      SoRu34980       5185
1      AkRu08112                 well      AkRu08112      59367
2      AkLu02044                river      AkLu02044      37379
3      AkHa00421                 well      AkHa00421      51627
4      SoRu35221                river      SoRu35221      28758

In [13]:
%%sql
SELECT 
ar.location_id AS audit_location,
ar.type_of_water_source,
v.location_id AS visit_location,
v.record_id,
wq.subjective_quality_score
FROM md_water_services.auditor_report AS ar
JOIN md_water_services.visits AS v
ON ar.location_id=v.location_id
JOIN water_quality AS wq
ON v.record_id=wq.record_id
LIMIT 5;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
5 rows affected.


audit_location type_of_water_source visit_location  record_id  \
0      SoRu34980                 well      SoRu34980       5185   
1      AkRu08112                 well      AkRu08112      59367   
2      AkLu02044                river      AkLu02044      37379   
3      AkHa00421                 well      AkHa00421      51627   
4      SoRu35221                river      SoRu35221      28758   

   subjective_quality_score  
0                         1  
1                         3  
2                         0  
3                         3  
4                         0

Dropping the duplicate Location column.
Since were joining 1620 rows of data, we want to keep track of the number of rows we get each time we run our query.

In [21]:
%%sql
SELECT 
ar.location_id,
v.record_id,
ar.true_water_source_score AS auditor_score,
wq.subjective_quality_score AS surveyor_score
FROM md_water_services.auditor_report AS ar
JOIN md_water_services.visits AS v
ON ar.location_id=v.location_id
JOIN water_quality AS wq
ON v.record_id=wq.record_id
LIMIT 10000;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2698 rows affected.


location_id  record_id  auditor_score  surveyor_score
0      SoRu34980       5185              1               1
1      AkRu08112      59367              3               3
2      AkLu02044      37379              0               0
3      AkHa00421      51627              3               3
4      SoRu35221      28758              0               0
...          ...        ...            ...             ...
2693   SoRu36378      19733              9               9
2694   KiRu27180      42498              6               6
2695   HaRu16981       2716              9               9
2696   HaDe16326      41884              1               1
2697   HaRu20394      24662              9               9

[2698 rows x 4 columns]

Checking if the auditor's and exployees' scores agree.

In [22]:
%%sql
SELECT 
ar.location_id,
v.record_id,
ar.true_water_source_score AS auditor_score,
wq.subjective_quality_score AS surveyor_score
FROM md_water_services.auditor_report AS ar
JOIN md_water_services.visits AS v
ON ar.location_id=v.location_id
JOIN water_quality AS wq
ON v.record_id=wq.record_id
WHERE ar.true_water_source_score=wq.subjective_quality_score
LIMIT 10000;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2505 rows affected.


location_id  record_id  auditor_score  surveyor_score
0      SoRu34980       5185              1               1
1      AkRu08112      59367              3               3
2      AkLu02044      37379              0               0
3      AkHa00421      51627              3               3
4      SoRu35221      28758              0               0
...          ...        ...            ...             ...
2500   SoRu36378      19733              9               9
2501   KiRu27180      42498              6               6
2502   HaRu16981       2716              9               9
2503   HaDe16326      41884              1               1
2504   HaRu20394      24662              9               9

[2505 rows x 4 columns]

The results show some of the locations were visited multiple times, so these records are duplicated here.

In [25]:
%%sql
SELECT 
ar.location_id,
v.record_id,
ar.true_water_source_score AS auditor_score,
wq.subjective_quality_score AS surveyor_score
FROM md_water_services.auditor_report AS ar
JOIN md_water_services.visits AS v
ON ar.location_id=v.location_id
JOIN water_quality AS wq
ON v.record_id=wq.record_id
WHERE ar.true_water_source_score=wq.subjective_quality_score
AND v.visit_count= 1
LIMIT 10000;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
1518 rows affected.


location_id  record_id  auditor_score  surveyor_score
0      SoRu34980       5185              1               1
1      AkRu08112      59367              3               3
2      AkLu02044      37379              0               0
3      AkHa00421      51627              3               3
4      SoRu35221      28758              0               0
...          ...        ...            ...             ...
1513   SoRu36378      19733              9               9
1514   KiRu27180      42498              6               6
1515   HaRu16981       2716              9               9
1516   HaDe16326      41884              1               1
1517   HaRu20394      24662              9               9

[1518 rows x 4 columns]

With the duplicates removed I we get 1518. 1518/1620 = 94% of the records the auditor checked were correct. Meaning that 102 records are incorrect.

In [27]:
%%sql
SELECT 
ar.location_id,
v.record_id,
ar.true_water_source_score AS auditor_score,
wq.subjective_quality_score AS surveyor_score
FROM md_water_services.auditor_report AS ar
JOIN md_water_services.visits AS v
ON ar.location_id=v.location_id
JOIN water_quality AS wq
ON v.record_id=wq.record_id
WHERE ar.true_water_source_score!=wq.subjective_quality_score
AND v.visit_count= 1
LIMIT 10000;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
102 rows affected.


location_id  record_id  auditor_score  surveyor_score
0     AkRu05215      21160              3              10
1     KiRu29290       7938              3              10
2     KiHa22748      43140              9              10
3     SoRu37841      18495              6              10
4     KiRu27884      33931              1              10
..          ...        ...            ...             ...
97    AmPw12480      47831              5              10
98    AmRu14842      47055              2              10
99    AkRu03358      31888              1              10
100   AmRu13433      57735              9              10
101   AmAm09956      15929              1              10

[102 rows x 4 columns]

In [31]:
%%sql
SELECT 
ar.location_id,
ar.type_of_water_source AS auditor_source,
ws.type_of_water_source AS survey_source,
v.record_id,
ar.true_water_source_score AS auditor_score,
wq.subjective_quality_score AS surveyor_score
FROM md_water_services.auditor_report AS ar
JOIN md_water_services.visits AS v
ON ar.location_id=v.location_id
JOIN water_quality AS wq
ON v.record_id=wq.record_id
JOIN water_source AS ws
ON v.source_id=ws.source_id
WHERE ar.true_water_source_score=wq.subjective_quality_score
LIMIT 10000;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
2505 rows affected.


location_id      auditor_source       survey_source  record_id  \
0      SoRu34980                well                well       5185   
1      AkRu08112                well                well      59367   
2      AkLu02044               river               river      37379   
3      AkHa00421                well                well      51627   
4      SoRu35221               river               river      28758   
...          ...                 ...                 ...        ...   
2500   SoRu36378  tap_in_home_broken  tap_in_home_broken      19733   
2501   KiRu27180          shared_tap          shared_tap      42498   
2502   HaRu16981  tap_in_home_broken  tap_in_home_broken       2716   
2503   HaDe16326                well                well      41884   
2504   HaRu20394  tap_in_home_broken  tap_in_home_broken      24662   

      auditor_score  surveyor_score  
0                 1               1  
1                 3               3  
2                 0               0  
3                 3               3  
4                 0               0  
...             ...             ...  
2500              9               9  
2501              6               6  
2502              9               9  
2503              1               1  
2504              9               9  

[2505 rows x 6 columns]

The types of sources look the same. Even though the scores are wrong, the integrity of the type_of_water_source data we analysed last time was not affected.

### Linking records to employees
At some of the locations, employees assigned scores incorrectly, and those records ended up in this results set.  
The are two reasons that can make this happen.  
1. These workers are all humans and make mistakes so this is expected.
2. Unfortunately, the alternative is that someone assigned scores incorrectly on purpose!

In [34]:
%%sql
SELECT 
ar.location_id,
v.record_id,
e.assigned_employee_id,
ar.true_water_source_score AS auditor_score,
wq.subjective_quality_score AS surveyor_score
FROM md_water_services.auditor_report AS ar
JOIN md_water_services.visits AS v
ON ar.location_id=v.location_id
JOIN water_quality AS wq
ON v.record_id=wq.record_id
JOIN employee AS e
ON v.assigned_employee_id=e.assigned_employee_id
WHERE ar.true_water_source_score!=wq.subjective_quality_score
AND v.visit_count= 1
LIMIT 10000;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
102 rows affected.


location_id  record_id  assigned_employee_id  auditor_score  \
0     AkRu05215      21160                    34              3   
1     KiRu29290       7938                     1              3   
2     KiHa22748      43140                     1              9   
3     SoRu37841      18495                    34              6   
4     KiRu27884      33931                     1              1   
..          ...        ...                   ...            ...   
97    AmPw12480      47831                     5              5   
98    AmRu14842      47055                     1              2   
99    AkRu03358      31888                     3              1   
100   AmRu13433      57735                     3              9   
101   AmAm09956      15929                     3              1   

     surveyor_score  
0                10  
1                10  
2                10  
3                10  
4                10  
..              ...  
97               10  
98               10  
99               10  
100              10  
101              10  

[102 rows x 5 columns]

We can link the incorrect records to the employees who recorded them, but the IDs don't help us identify them. We have employees' names
stored along with their IDs, so let's fetch their names from the employees table instead of the IDs.

In [38]:
%%sql
WITH Incorrect_records AS(
SELECT 
ar.location_id,
v.record_id,
e.employee_name,
ar.true_water_source_score AS auditor_score,
wq.subjective_quality_score AS surveyor_score
FROM md_water_services.auditor_report AS ar
JOIN md_water_services.visits AS v
ON ar.location_id=v.location_id
JOIN water_quality AS wq
ON v.record_id=wq.record_id
JOIN employee AS e
ON v.assigned_employee_id=e.assigned_employee_id
WHERE ar.true_water_source_score!=wq.subjective_quality_score
AND v.visit_count= 1
LIMIT 10000)

SELECT * FROM Incorrect_records;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
102 rows affected.


location_id  record_id   employee_name  auditor_score  surveyor_score
0     AkRu05215      21160      Rudo Imani              3              10
1     KiRu29290       7938     Bello Azibo              3              10
2     KiHa22748      43140     Bello Azibo              9              10
3     SoRu37841      18495      Rudo Imani              6              10
4     KiRu27884      33931     Bello Azibo              1              10
..          ...        ...             ...            ...             ...
97    AmPw12480      47831  Zuriel Matembo              5              10
98    AmRu14842      47055     Bello Azibo              2              10
99    AkRu03358      31888  Malachi Mavuso              1              10
100   AmRu13433      57735  Malachi Mavuso              9              10
101   AmAm09956      15929  Malachi Mavuso              1              10

[102 rows x 5 columns]

Unique list of employees from the above table

In [44]:
%%sql
WITH Incorrect_records AS(
SELECT 
ar.location_id,
v.record_id,
e.employee_name,
ar.true_water_source_score AS auditor_score,
wq.subjective_quality_score AS surveyor_score
FROM md_water_services.auditor_report AS ar
JOIN md_water_services.visits AS v
ON ar.location_id=v.location_id
JOIN water_quality AS wq
ON v.record_id=wq.record_id
JOIN employee AS e
ON v.assigned_employee_id=e.assigned_employee_id
WHERE ar.true_water_source_score!=wq.subjective_quality_score
AND v.visit_count= 1
LIMIT 10000)

SELECT 
DISTINCT employee_name,
COUNT(employee_name) AS number_of_mistakes
FROM Incorrect_records
GROUP BY employee_name
ORDER BY number_of_mistakes DESC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
17 rows affected.


employee_name  number_of_mistakes
0      Bello Azibo                  26
1   Malachi Mavuso                  21
2   Zuriel Matembo                  17
3   Lalitha Kaburi                   7
4       Rudo Imani                   5
5        Farai Nia                   4
6      Enitan Zuri                   4
7    Yewande Ebele                   3
8    Jengo Tumaini                   3
9     Makena Thabo                   3
10     Gamba Shani                   3
11   Thandiwe Kito                   1
12       Pili Zola                   1
13       Usafi Ayo                   1
14    Deka Osumare                   1
15        Ona Sefu                   1
16      Xola Uzuri                   1

Some of the surveyors were making a lot of mistakes while many of the other surveyors were only making a few.

## Gathering Evidence
How would we go about finding out if any of our employees are corrupt?  
Let's say all employees make mistakes; if someone is corrupt, they will be making a lot of "mistakes", more than average, for example. But someone
could just be clumsy, so we should try to get more evidence.
#### So let's try to find all of the employees who have an above-average number of mistakes. 
We have to compare each employee's error_count with avg_error_count_per_empl.

In [51]:
%%sql
WITH Incorrect_records AS(
SELECT 
ar.location_id,
v.record_id,
e.employee_name,
ar.true_water_source_score AS auditor_score,
wq.subjective_quality_score AS surveyor_score
FROM md_water_services.auditor_report AS ar
JOIN md_water_services.visits AS v
ON ar.location_id=v.location_id
JOIN water_quality AS wq
ON v.record_id=wq.record_id
JOIN employee AS e
ON v.assigned_employee_id=e.assigned_employee_id
WHERE ar.true_water_source_score!=wq.subjective_quality_score
AND v.visit_count= 1
LIMIT 10000),

error_count AS(
SELECT 
DISTINCT employee_name,
COUNT(employee_name) AS number_of_mistakes
FROM Incorrect_records
GROUP BY employee_name
ORDER BY number_of_mistakes DESC)

SELECT
AVG(number_of_mistakes) AS avg_error_count_per_empl
FROM error_count;



 * mysql+pymysql://root:***@localhost:3306/md_water_services
1 rows affected.


avg_error_count_per_empl
0                   6.0000

In [5]:
%%sql
CREATE VIEW Incorrect_records AS(
SELECT 
ar.location_id,
v.record_id,
e.employee_name,
ar.true_water_source_score AS auditor_score,
wq.subjective_quality_score AS surveyor_score,
ar.statements AS statements
FROM md_water_services.auditor_report AS ar
JOIN 
   md_water_services.visits AS v
ON 
  ar.location_id=v.location_id
JOIN 
   water_quality AS wq
ON 
  v.record_id=wq.record_id
JOIN 
   employee AS e
ON 
  v.assigned_employee_id=e.assigned_employee_id
WHERE 
    ar.true_water_source_score!=wq.subjective_quality_score
AND v.visit_count= 1
LIMIT 10000);

 * mysql+pymysql://root:***@localhost:3306/md_water_services
0 rows affected.


Empty DataFrame
Columns: []
Index: []

In [9]:
%%sql
WITH error_count AS(
SELECT 
DISTINCT employee_name,
COUNT(employee_name) AS number_of_mistakes
FROM Incorrect_records
GROUP BY employee_name
ORDER BY number_of_mistakes DESC
)
SELECT * 
FROM error_count;


 * mysql+pymysql://root:***@localhost:3306/md_water_services
17 rows affected.


employee_name  number_of_mistakes
0      Bello Azibo                  26
1   Malachi Mavuso                  21
2   Zuriel Matembo                  17
3   Lalitha Kaburi                   7
4       Rudo Imani                   5
5        Farai Nia                   4
6      Enitan Zuri                   4
7    Yewande Ebele                   3
8    Jengo Tumaini                   3
9     Makena Thabo                   3
10     Gamba Shani                   3
11   Thandiwe Kito                   1
12       Pili Zola                   1
13       Usafi Ayo                   1
14    Deka Osumare                   1
15        Ona Sefu                   1
16      Xola Uzuri                   1

In [21]:
%%sql
DROP VIEW error_count;
CREATE VIEW error_count AS(
SELECT 
DISTINCT employee_name,
COUNT(employee_name) AS number_of_mistakes
FROM Incorrect_records
GROUP BY employee_name
ORDER BY number_of_mistakes DESC
);

 * mysql+pymysql://root:***@localhost:3306/md_water_services
0 rows affected.
0 rows affected.


Empty DataFrame
Columns: []
Index: []

## Employees who made more mistakes than the average person,

In [23]:
%%sql
SELECT
employee_name,
number_of_mistakes
FROM error_count
WHERE number_of_mistakes>(SELECT
                           AVG(number_of_mistakes) AS avg_error_count_per_empl
                            FROM error_count);

 * mysql+pymysql://root:***@localhost:3306/md_water_services
4 rows affected.


employee_name  number_of_mistakes
0     Bello Azibo                  26
1  Malachi Mavuso                  21
2  Zuriel Matembo                  17
3  Lalitha Kaburi                   7

#### Suspect List

In [28]:
%%sql
CREATE VIEW suspect_list AS (SELECT
employee_name,
number_of_mistakes
FROM error_count
WHERE number_of_mistakes>(SELECT
                           AVG(number_of_mistakes) AS avg_error_count_per_empl
                            FROM error_count));
SELECT employee_name
FROM suspect_list;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
0 rows affected.
4 rows affected.


employee_name
0     Bello Azibo
1  Malachi Mavuso
2  Zuriel Matembo
3  Lalitha Kaburi

Filtering the Incorrect_records view to identify all of the records associated with the four employees we identified.

In [31]:
%%sql
SELECT 
employee_name,
location_id,
statements
FROM Incorrect_records
WHERE employee_name IN(
    SELECT employee_name
    FROM suspect_list)
ORDER BY employee_name ASC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
71 rows affected.


employee_name location_id  \
0      Bello Azibo   AmRu14691   
1      Bello Azibo   KiRu29290   
2      Bello Azibo   KiHa22748   
3      Bello Azibo   KiZu30847   
4      Bello Azibo   KiRu27884   
..             ...         ...   
66  Zuriel Matembo   SoRu38331   
67  Zuriel Matembo   AmPw12480   
68  Zuriel Matembo   HaSe20888   
69  Zuriel Matembo   SoRu38401   
70  Zuriel Matembo   KiZu31170   

                                           statements  
0   An artisan's creative mind is stifled by the w...  
1   A young artist sketches the faces in the queue...  
2   A young girl's hopeful eyes are clouded by mis...  
3   A traveling trader's pause in the village incl...  
4   A traditional healer's empathy turns to bitter...  
..                                                ...  
66  An unsettling atmosphere surrounded the offici...  
67  A village chief's dignity is marred by frustra...  
68  Suspicion and unease colored the villagers' ac...  
69  A village chief's dignified wait speaks of lea...  
70  A community leader stood with his people, expr...  

[71 rows x 3 columns]

We notice some alarming statements about these four officials (these records: AkRu04508, AkRu07310,
KiRu29639, AmAm09607, for example, we see how the word cash is used a lot below.

In [33]:
%%sql
SELECT 
employee_name,
location_id,
statements
FROM Incorrect_records
WHERE employee_name IN(
    SELECT employee_name
    FROM suspect_list) AND statements LIKE'%cash%'
ORDER BY employee_name ASC;

 * mysql+pymysql://root:***@localhost:3306/md_water_services
19 rows affected.


employee_name location_id  \
0      Bello Azibo   KiIs23853   
1      Bello Azibo   HaSe21323   
2      Bello Azibo   KiRu27065   
3      Bello Azibo   AkRu04508   
4      Bello Azibo   KiRu29639   
5      Bello Azibo   KiMr24919   
6      Bello Azibo   AmRu14887   
7      Bello Azibo   AkRu05741   
8   Lalitha Kaburi   AkRu07310   
9   Lalitha Kaburi   KiRu29329   
10  Malachi Mavuso   AmAm09607   
11  Malachi Mavuso   KiRu25347   
12  Malachi Mavuso   AmRu15719   
13  Malachi Mavuso   AmAm09956   
14  Zuriel Matembo   SoRu38331   
15  Zuriel Matembo   AkRu05880   
16  Zuriel Matembo   SoIl32575   
17  Zuriel Matembo   HaSe20888   
18  Zuriel Matembo   HaRu20146   

                                           statements  
0   Villagers' wary accounts of an official's arro...  
1   Villagers spoke of an unsettling encounter wit...  
2   Villagers expressed their discomfort with an o...  
3   An unsettling atmosphere surrounded the offici...  
4   An unsettling atmosphere prevailed as villager...  
5   Suspicion and unease colored the villagers' ac...  
6   Villagers expressed their discomfort with an o...  
7   An air of mistrust surrounded the official, as...  
8   Villagers spoke of their unsettling encounters...  
9   Suspicion colored villagers' descriptions of a...  
10  Villagers spoke of an unsettling encounter wit...  
11  Villagers expressed their discontent with an o...  
12  Suspicion and unease colored the villagers' ac...  
13  An unsettling atmosphere surrounded the offici...  
14  An unsettling atmosphere surrounded the offici...  
15  Villagers' wary accounts of an official's arro...  
16  Villagers recounted unsettling encounters with...  
17  Suspicion and unease colored the villagers' ac...  
18  An air of mistrust hung heavy as villagers des...

So we can sum up the evidence we have for Zuriel Matembo, Malachi Mavuso, Bello Azibo and Lalitha Kaburi:  
1. They all made more mistakes than their peers on average.  
2. They all have incriminating statements made against them, and only them.  
This is not decisive proof, but it is concerning enough that we should flag it.